File: src/main/java/com/reactive/nexo/siau/service/AsignacionService.java
package com.reactive.nexo.siau.service;

import com.reactive.nexo.siau.dto.AsignacionResponse;
import com.reactive.nexo.siau.dto.CreateAsignacionRequest;
import com.reactive.nexo.siau.dto.DashboardStatsResponse;
import com.reactive.nexo.siau.entity.AsignacionEntity;
import com.reactive.nexo.siau.entity.DepartamentoEntity;
import com.reactive.nexo.siau.entity.PqrsdfEntity;
import com.reactive.nexo.siau.model.EstadoAsignacion;
import com.reactive.nexo.siau.model.EstadoPqrsdf;
import com.reactive.nexo.siau.repository.AsignacionRepository;
import com.reactive.nexo.siau.repository.DepartamentoRepository;
import com.reactive.nexo.siau.repository.PqrsdfRepository;
import com.reactive.nexo.siau.repository.TrazabilidadRepository;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import reactor.core.scheduler.Schedulers;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.*;

@Service
public class AsignacionService {

    private static final Logger log = LoggerFactory.getLogger(AsignacionService.class);

    private final AsignacionRepository repo;
    private final PqrsdfRepository pqrsdfRepo;
    private final DepartamentoRepository departamentoRepo;
    private final TrazabilidadRepository trazabilidadRepo;
    private final PqrsdfService pqrsdfService;

    public AsignacionService(AsignacionRepository repo,
                             PqrsdfRepository pqrsdfRepo,
                             DepartamentoRepository departamentoRepo,
                             TrazabilidadRepository trazabilidadRepo,
                             PqrsdfService pqrsdfService) {
        this.repo = repo;
        this.pqrsdfRepo = pqrsdfRepo;
        this.departamentoRepo = departamentoRepo;
        this.trazabilidadRepo = trazabilidadRepo;
        this.pqrsdfService = pqrsdfService;
    }

    public AsignacionResponse create(CreateAsignacionRequest req, Long employeeId) {
        // Validar que la PQRSDF existe
        PqrsdfEntity pqrsdf = pqrsdfRepo.findById(req.getPqrsdfId())
                .subscribeOn(Schedulers.boundedElastic())
                .block();
        if (pqrsdf == null) {
            throw new IllegalArgumentException("PQRSDF con ID " + req.getPqrsdfId() + " no encontrada");
        }

        // Validar que el departamento existe
        DepartamentoEntity depto = departamentoRepo.findById(req.getDepartamentoId())
                .subscribeOn(Schedulers.boundedElastic())
                .block();
        if (depto == null) {
            throw new IllegalArgumentException("Departamento con ID " + req.getDepartamentoId() + " no encontrado");
        }

        // Calcular fecha límite por defecto (15 días hábiles si no se especifica)
        LocalDate fechaLimite = req.getFechaLimiteRespuesta() != null
                ? req.getFechaLimiteRespuesta()
                : calcularFechaLimite(LocalDate.now(), 15);

        AsignacionEntity entity = AsignacionEntity.builder()
                .pqrsdfId(req.getPqrsdfId())
                .departamentoId(req.getDepartamentoId())
                .funcionarioId(req.getFuncionarioId())
                .funcionarioNombre(req.getFuncionarioNombre())
                .fechaAsignacion(LocalDateTime.now())
                .fechaLimiteRespuesta(fechaLimite)
                .estado(EstadoAsignacion.PENDIENTE.getValue())
                .observaciones(req.getObservaciones())
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();

        AsignacionEntity saved = repo.save(entity)
                .subscribeOn(Schedulers.boundedElastic())
                .block();

        // Actualizar estado de la PQRSDF a ASIGNADO
        pqrsdf.setEstado(EstadoPqrsdf.ASIGNADO.getValue());
        pqrsdf.setUpdatedAt(LocalDateTime.now());
        pqrsdfRepo.save(pqrsdf)
                .subscribeOn(Schedulers.boundedElastic())
                .subscribe();

        // Registrar trazabilidad
        pqrsdfService.getTrazabilidad(req.getPqrsdfId()); // Ensure method access - actually we call repo directly
        registrarTrazabilidadInterna(req.getPqrsdfId(), "ASIGNACION",
                "Asignada a: " + depto.getNombre() + " - Fecha límite: " + fechaLimite,
                employeeId, null);

        log.info("Asignación creada: PQRSDF {} → Departamento {} (ID: {})",
                pqrsdf.getConsecutivo(), depto.getNombre(), saved.getId());

        return mapToResponse(saved, pqrsdf.getConsecutivo(), depto.getNombre());
    }

    public Optional<AsignacionResponse> findById(Long id) {
        AsignacionEntity entity = repo.findById(id)
                .subscribeOn(Schedulers.boundedElastic())
                .block();
        if (entity == null) return Optional.empty();

        PqrsdfEntity pqrsdf = pqrsdfRepo.findById(entity.getPqrsdfId())
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(null);
        DepartamentoEntity depto = departamentoRepo.findById(entity.getDepartamentoId())
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(null);

        return Optional.of(mapToResponse(entity,
                pqrsdf != null ? pqrsdf.getConsecutivo() : null,
                depto != null ? depto.getNombre() : null));
    }

    public List<AsignacionResponse> findByPqrsdfId(Long pqrsdfId) {
        List<AsignacionEntity> list = repo.findByPqrsdfId(pqrsdfId)
                .collectList()
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(Collections.emptyList());

        List<AsignacionResponse> result = new ArrayList<>();
        for (AsignacionEntity e : list) {
            PqrsdfEntity p = pqrsdfRepo.findById(e.getPqrsdfId())
                    .subscribeOn(Schedulers.boundedElastic())
                    .blockOptional().orElse(null);
            DepartamentoEntity d = departamentoRepo.findById(e.getDepartamentoId())
                    .subscribeOn(Schedulers.boundedElastic())
                    .blockOptional().orElse(null);
            result.add(mapToResponse(e,
                    p != null ? p.getConsecutivo() : null,
                    d != null ? d.getNombre() : null));
        }
        return result;
    }

    public List<AsignacionResponse> findByDepartamentoId(Long departamentoId) {
        List<AsignacionEntity> list = repo.findByDepartamentoId(departamentoId)
                .collectList()
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(Collections.emptyList());

        List<AsignacionResponse> result = new ArrayList<>();
        for (AsignacionEntity e : list) {
            PqrsdfEntity p = pqrsdfRepo.findById(e.getPqrsdfId())
                    .subscribeOn(Schedulers.boundedElastic())
                    .blockOptional().orElse(null);
            DepartamentoEntity d = departamentoRepo.findById(e.getDepartamentoId())
                    .subscribeOn(Schedulers.boundedElastic())
                    .blockOptional().orElse(null);
            result.add(mapToResponse(e,
                    p != null ? p.getConsecutivo() : null,
                    d != null ? d.getNombre() : null));
        }
        return result;
    }

    public AsignacionResponse responderAsignacion(Long id, String respuesta, Long employeeId) {
        AsignacionEntity entity = repo.findById(id)
                .subscribeOn(Schedulers.boundedElastic())
                .block();
        if (entity == null) return null;

        entity.setRespuestaArea(respuesta);
        entity.setEstado(EstadoAsignacion.RESPONDIDA.getValue());
        entity.setFechaRespuestaArea(LocalDateTime.now());
        entity.setUpdatedAt(LocalDateTime.now());

        AsignacionEntity saved = repo.save(entity)
                .subscribeOn(Schedulers.boundedElastic())
                .block();

        PqrsdfEntity pqrsdf = pqrsdfRepo.findById(entity.getPqrsdfId())
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(null);
        DepartamentoEntity depto = departamentoRepo.findById(entity.getDepartamentoId())
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(null);

        String consecutivo = pqrsdf != null ? pqrsdf.getConsecutivo() : null;
        String deptoNombre = depto != null ? depto.getNombre() : null;

        registrarTrazabilidadInterna(entity.getPqrsdfId(), "RESPUESTA_AREA",
                "Departamento " + deptoNombre + " respondió asignación",
                employeeId, null);

        log.info("Asignación {} respondida por departamento {}", id, deptoNombre);
        return mapToResponse(saved, consecutivo, deptoNombre);
    }

    public DashboardStatsResponse getDashboardStats() {
        LocalDate hoy = LocalDate.now();
        LocalDate hace30dias = hoy.minusDays(30);

        long total = pqrsdfService.count(new HashMap<>());
        long pendientes = pqrsdfRepo.countByEstado("RECIBIDO")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        long enGestion = pqrsdfRepo.countByEstado("ASIGNADO")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        long enGestion2 = pqrsdfRepo.countByEstado("EN_GESTION")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        long respondidas = pqrsdfRepo.countByEstado("RESPONDIDO")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        long cerradas = pqrsdfRepo.countByEstado("CERRADO")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);

        long asignacionesVencidas = repo.countByEstado("VENCIDA")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);

        long porTipoP = pqrsdfRepo.countByTipo("P")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        long porTipoQ = pqrsdfRepo.countByTipo("Q")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        long porTipoR = pqrsdfRepo.countByTipo("R")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        long porTipoS = pqrsdfRepo.countByTipo("S")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        long porTipoD = pqrsdfRepo.countByTipo("D")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        long porTipoF = pqrsdfRepo.countByTipo("F")
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);

        Map<String, Long> porTipo = new LinkedHashMap<>();
        porTipo.put("PETICION", porTipoP);
        porTipo.put("QUEJA", porTipoQ);
        porTipo.put("RECLAMO", porTipoR);
        porTipo.put("SUGERENCIA", porTipoS);
        porTipo.put("DENUNCIA", porTipoD);
        porTipo.put("FELICITACION", porTipoF);

        // Stats por departamento
        List<DepartamentoEntity> deptos = departamentoRepo.findAll()
                .collectList()
                .subscribeOn(Schedulers.boundedElastic())
                .blockOptional().orElse(Collections.emptyList());
        Map<String, Long> porDepartamento = new LinkedHashMap<>();
        for (DepartamentoEntity d : deptos) {
            Long count = repo.countByDepartamentoAndEstado(d.getId(), "PENDIENTE")
                    .subscribeOn(Schedulers.boundedElastic())
                    .blockOptional().orElse(0L);
            if (count > 0) porDepartamento.put(d.getNombre(), count);
        }

        // Últimos 30 días
        long ultimos30 = pqrsdfRepo.countTotal()
                .subscribeOn(Schedulers.boundedElastic()).blockOptional().orElse(0L);
        Map<String, Long> ultimos30Map = new HashMap<>();
        ultimos30Map.put("total30dias", ultimos30 > 0 ? ultimos30 : 0);

        return DashboardStatsResponse.builder()
                .totalPqrsdf(total)
                .pendientes(pendientes + enGestion + enGestion2)
                .enGestion(enGestion + enGestion2)
                .respondidas(respondidas)
                .cerradas(cerradas)
                .vencidas(asignacionesVencidas)
                .porTipo(porTipo)
                .porDepartamento(porDepartamento)
                .ultimos30Dias(ultimos30Map)
                .build();
    }

    // --- Métodos auxiliares ---

    private void registrarTrazabilidadInterna(Long pqrsdfId, String accion, String descripcion,
                                               Long usuarioId, String metadata) {
        com.reactive.nexo.siau.entity.TrazabilidadEntity entity =
                com.reactive.nexo.siau.entity.TrazabilidadEntity.builder()
                        .pqrsdfId(pqrsdfId)
                        .accion(accion)
                        .descripcion(descripcion)
                        .usuarioId(usuarioId)
                        .createdAt(LocalDateTime.now())
                        .build();
        trazabilidadRepo.save(entity)
                .subscribeOn(Schedulers.boundedElastic())
                .subscribe();
    }

    private AsignacionResponse mapToResponse(AsignacionEntity e, String consecutivo, String deptoNombre) {
        boolean vencida = e.getFechaLimiteRespuesta() != null
                && e.getFechaLimiteRespuesta().isBefore(LocalDate.now())
                && !"RESPONDIDA".equals(e.getEstado());

        int diasRestantes = e.getFechaLimiteRespuesta() != null
                ? (int) ChronoUnit.DAYS.between(LocalDate.now(), e.getFechaLimiteRespuesta())
                : 0;

        return AsignacionResponse.builder()
                .id(e.getId())
                .pqrsdfId(e.getPqrsdfId())
                .pqrsdfConsecutivo(consecutivo)
                .departamentoId(e.getDepartamentoId())
                .departamentoNombre(deptoNombre)
                .funcionarioId(e.getFuncionarioId())
                .funcionarioNombre(e.getFuncionarioNombre())
                .fechaAsignacion(e.getFechaAsignacion())
                .fechaLimiteRespuesta(e.getFechaLimiteRespuesta())
                .estado(e.getEstado())
                .respuestaArea(e.getRespuestaArea())
                .observaciones(e.getObservaciones())
                .fechaRespuestaArea(e.getFechaRespuestaArea())
                .createdAt(e.getCreatedAt())
                .vencida(vencida)
                .diasRestantes(diasRestantes)
                .build();
    }

    /**
     * Calcula una fecha límite sumando días hábiles (sin contar sábados ni domingos).
     */
    private LocalDate calcularFechaLimite(LocalDate desde, int diasHabiles) {
        LocalDate result = desde;
        int agregados = 0;
        while (agregados < diasHabiles) {
            result = result.plusDays(1);
            if (result.getDayOfWeek() != DayOfWeek.SATURDAY && result.getDayOfWeek() != DayOfWeek.SUNDAY) {
                agregados++;
            }
        }
        return result;
    }
}

File: app/services/asignacion_service.py
"""Service layer for Asignacion — assignment of PQRSDF to departments."""

from __future__ import annotations

from datetime import date, datetime, timedelta, timezone

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func

from app.repositories import asignacion_repo, pqrsdf_repo, departamento_repo, trazabilidad_repo
from app.schemas import (
    CreateAsignacionRequest,
    AsignacionResponse,
    DashboardStatsResponse,
)
from app.models import Asignacion, Pqrsdf, Trazabilidad


# ──────────────────────────────────────────────
# Helpers
# ──────────────────────────────────────────────

def _calcular_fecha_limite(desde: date, dias_habiles: int = 15) -> date:
    """Calcula una fecha límite sumando días hábiles (salta sábados y domingos)."""
    dias_agregados = 0
    fecha = desde
    while dias_agregados < dias_habiles:
        fecha += timedelta(days=1)
        if fecha.weekday() < 5:  # 0=Lun ... 4=Vie
            dias_agregados += 1
    return fecha


def _map_to_response(
    asignacion: Asignacion,
    consecutivo: str | None = None,
    depto_nombre: str | None = None,
) -> AsignacionResponse:
    """Mapea entidad Asignacion a AsignacionResponse."""
    vencida = False
    dias_restantes = 0
    if (
        asignacion.estado
        and asignacion.estado not in ("RESPONDIDA", "CERRADA")
        and asignacion.fecha_limite_respuesta
    ):
        hoy = date.today()
        dias_restantes = (asignacion.fecha_limite_respuesta - hoy).days
        vencida = dias_restantes < 0
        if vencida:
            dias_restantes = abs(dias_restantes)

    return AsignacionResponse(
        id=asignacion.id,
        pqrsdfId=asignacion.pqrsdf_id,
        pqrsdfConsecutivo=consecutivo,
        departamentoId=asignacion.departamento_id,
        departamentoNombre=depto_nombre or (
            asignacion.departamento_rel.nombre
            if asignacion.departamento_rel
            else None
        ),
        funcionarioId=asignacion.funcionario_id,
        funcionarioNombre=asignacion.funcionario_nombre,
        fechaAsignacion=asignacion.fecha_asignacion,
        fechaLimiteRespuesta=asignacion.fecha_limite_respuesta,
        estado=asignacion.estado,
        respuestaArea=asignacion.respuesta_area,
        observaciones=asignacion.observaciones,
        fechaRespuestaArea=asignacion.fecha_respuesta_area,
        createdAt=asignacion.created_at,
        vencida=vencida,
        diasRestantes=dias_restantes,
    )


def _get_departamento_nombre(asignacion: Asignacion) -> str | None:
    """Obtiene nombre del departamento desde la relación cargada."""
    if asignacion.departamento_rel:
        return asignacion.departamento_rel.nombre
    return None


# ──────────────────────────────────────────────
# CRUD
# ──────────────────────────────────────────────

async def create(
    session: AsyncSession,
    req: CreateAsignacionRequest,
    employee_id: int,
) -> AsignacionResponse:
    """Crea una nueva asignación y actualiza estado de PQRSDF a ASIGNADO."""
    # Validar que PQRSDF exista
    pqrsdf = await pqrsdf_repo.find_by_id(session, req.pqrsdfId)
    if pqrsdf is None:
        raise ValueError(f"PQRSDF con ID {req.pqrsdfId} no encontrada")

    # Validar que departamento exista
    depto = await departamento_repo.find_by_id(session, req.departamentoId)
    if depto is None:
        raise ValueError(f"Departamento con ID {req.departamentoId} no encontrado")

    # Calcular fecha límite
    fecha_limite = req.fechaLimiteRespuesta or _calcular_fecha_limite(date.today())

    entity = Asignacion(
        pqrsdf_id=req.pqrsdfId,
        departamento_id=req.departamentoId,
        funcionario_id=req.funcionarioId,
        funcionario_nombre=req.funcionarioNombre,
        fecha_asignacion=datetime.now(timezone.utc),
        fecha_limite_respuesta=fecha_limite,
        estado="PENDIENTE",
        observaciones=req.observaciones,
        created_at=datetime.now(timezone.utc),
        updated_at=datetime.now(timezone.utc),
    )

    saved = await asignacion_repo.save(session, entity)

    # Actualizar estado de PQRSDF
    if pqrsdf.estado in ("RECIBIDO", "PENDIENTE"):
        pqrsdf.estado = "ASIGNADO"
        pqrsdf.updated_by = employee_id
        pqrsdf.updated_at = datetime.now(timezone.utc)
        await pqrsdf_repo.save(session, pqrsdf)

    # Registrar trazabilidad
    trazabilidad = Trazabilidad(
        pqrsdf_id=req.pqrsdfId,
        accion="ASIGNACION",
        descripcion=f"Asignada a {depto.nombre} — vence {fecha_limite.isoformat()}",
        usuario_id=employee_id,
        created_at=datetime.now(timezone.utc),
    )
    await trazabilidad_repo.save(session, trazabilidad)

    return _map_to_response(
        saved,
        consecutivo=pqrsdf.consecutivo,
        depto_nombre=depto.nombre,
    )


async def get_by_id(
    session: AsyncSession,
    id: int,
) -> AsignacionResponse | None:
    """Alias router-compatible para find_by_id."""
    return await find_by_id(session, id)


async def find_by_id(
    session: AsyncSession,
    id: int,
) -> AsignacionResponse | None:
    """Busca una asignación por ID."""
    entity = await asignacion_repo.find_by_id(session, id)
    if entity is None:
        return None
    return _map_to_response(entity)


async def list_by_pqrsdf(
    session: AsyncSession,
    pqrsdf_id: int,
) -> list[AsignacionResponse]:
    """Alias router-compatible para find_by_pqrsdf_id."""
    return await find_by_pqrsdf_id(session, pqrsdf_id)


async def find_by_pqrsdf_id(
    session: AsyncSession,
    pqrsdf_id: int,
) -> list[AsignacionResponse]:
    """Retorna todas las asignaciones de una PQRSDF."""
    entities = await asignacion_repo.find_by_pqrsdf_id(session, pqrsdf_id)
    pqrsdf = await pqrsdf_repo.find_by_id(session, pqrsdf_id)
    consecutivo = pqrsdf.consecutivo if pqrsdf else None

    responses = []
    for e in entities:
        depto_nombre = _get_departamento_nombre(e)
        responses.append(_map_to_response(e, consecutivo=consecutivo, depto_nombre=depto_nombre))
    return responses


async def list_by_departamento(
    session: AsyncSession,
    departamento_id: int,
) -> list[AsignacionResponse]:
    """Alias router-compatible para find_by_departamento_id."""
    return await find_by_departamento_id(session, departamento_id)


async def find_by_departamento_id(
    session: AsyncSession,
    departamento_id: int,
) -> list[AsignacionResponse]:
    """Retorna asignaciones activas de un departamento."""
    entities = await asignacion_repo.find_by_departamento_id(session, departamento_id)

    responses = []
    for e in entities:
        consecutivo = e.pqrsdf_rel.consecutivo if e.pqrsdf_rel else None
        depto_nombre = _get_departamento_nombre(e)
        responses.append(_map_to_response(e, consecutivo=consecutivo, depto_nombre=depto_nombre))
    return responses


async def responder(
    session: AsyncSession,
    id: int,
    body: dict,
    employee_id: int,
) -> AsignacionResponse | None:
    """Alias router-compatible para responder_asignacion (body es dict con campo 'respuesta')."""
    return await responder_asignacion(session, id, body.get("respuesta", ""), employee_id)


async def responder_asignacion(
    session: AsyncSession,
    id: int,
    respuesta: str,
    employee_id: int,
) -> AsignacionResponse | None:
    """Registra respuesta del área sobre una asignación."""
    entity = await asignacion_repo.find_by_id(session, id)
    if entity is None:
        return None

    entity.respuesta_area = respuesta
    entity.estado = "RESPONDIDA"
    entity.fecha_respuesta_area = datetime.now(timezone.utc)
    entity.updated_at = datetime.now(timezone.utc)

    await asignacion_repo.save(session, entity)

    # Registrar trazabilidad
    pqrsdf = await pqrsdf_repo.find_by_id(session, entity.pqrsdf_id)
    depto_nombre = entity.departamento_rel.nombre if entity.departamento_rel else "desconocido"

    trazabilidad = Trazabilidad(
        pqrsdf_id=entity.pqrsdf_id,
        accion="RESPUESTA_AREA",
        descripcion=f"Respuesta recibida del área {depto_nombre}",
        usuario_id=employee_id,
        created_at=datetime.now(timezone.utc),
    )
    await trazabilidad_repo.save(session, trazabilidad)

    return _map_to_response(
        entity,
        consecutivo=pqrsdf.consecutivo if pqrsdf else None,
        depto_nombre=depto_nombre,
    )


async def get_dashboard_stats(
    session: AsyncSession,
) -> DashboardStatsResponse:
    """Calcula estadísticas para el dashboard: conteo por estado, tipo y departamento."""

    # Totales por estado
    stmt_estados = select(Pqrsdf.estado, func.count(Pqrsdf.id).label("total")).group_by(Pqrsdf.estado)
    result = await session.execute(stmt_estados)
    estado_counts = dict(result.all())

    # Totales por tipo
    stmt_tipos = select(Pqrsdf.tipo, func.count(Pqrsdf.id).label("total")).group_by(Pqrsdf.tipo)
    result = await session.execute(stmt_tipos)
    tipo_counts = dict(result.all())

    # Totales por departamento (asignaciones activas)
    stmt_deptos = select(
        Asignacion.departamento_id, func.count(Asignacion.id).label("total")
    ).where(
        Asignacion.estado.in_(["PENDIENTE", "EN_PROCESO"])
    ).group_by(Asignacion.departamento_id)
    result = await session.execute(stmt_deptos)
    depto_counts_raw = dict(result.all())

    # Mapear IDs de departamento a nombres
    depto_counts: dict[str, int] = {}
    for depto_id, count in depto_counts_raw.items():
        depto = await departamento_repo.find_by_id(session, depto_id)
        nombre = depto.nombre if depto else f"ID {depto_id}"
        depto_counts[nombre] = count

    # Últimos 30 días
    hace_30 = date.today() - timedelta(days=30)
    stmt_30d = select(
        func.date(Pqrsdf.created_at).label("dia"),
        func.count(Pqrsdf.id).label("total"),
    ).where(
        Pqrsdf.created_at >= hace_30
    ).group_by(
        func.date(Pqrsdf.created_at)
    ).order_by(
        func.date(Pqrsdf.created_at)
    )
    result = await session.execute(stmt_30d)
    ultimos_30dias = {str(row.dia): row.total for row in result}

    total = sum(estado_counts.values())
    pendientes = estado_counts.get("RECIBIDO", 0) + estado_counts.get("PENDIENTE", 0)
    en_gestion = estado_counts.get("ASIGNADO", 0) + estado_counts.get("EN_GESTION", 0)
    respondidas = estado_counts.get("RESPONDIDO", 0)

    # Vencidas: asignaciones con fecha_limite_respuesta < today y estado activo
    stmt_vencidas = select(func.count(Asignacion.id)).where(
        Asignacion.fecha_limite_respuesta < date.today(),
        Asignacion.estado.in_(["PENDIENTE", "EN_PROCESO"]),
    )
    result = await session.execute(stmt_vencidas)
    vencidas = result.scalar() or 0

    return DashboardStatsResponse(
        totalPqrsdf=total,
        pendientes=pendientes,
        enGestion=en_gestion,
        respondidas=respondidas,
        cerradas=estado_counts.get("CERRADO", 0),
        vencidas=vencidas,
        porTipo=dict(tipo_counts),
        porDepartamento=depto_counts,
        ultimos30Dias=ultimos_30dias,
    )
